proposals/devcontainer-features.md

```diff
@@ -0,0 +1,143 @@
+# Dev container features reference
+
+Dev container features provide a smooth path for customizing your container definitions. Features are self contained units of code meant to facilitate creation of custom images for application or development containers.
+
+From a practical point of view, features are folders that contain units of code with different entrypoints for different lifecycle events.
+
+Features can be defined by a `feature.json` file in the root folder of the feature. The file is optional for backwards compatibility but it is required for any new features being authored.
+
+Features are to be executed in sequence as defined in `devcontainer.json`.
+
+## Folder Structure 
+
+A feature is a self contained entity in a folder. A feature release would be a tar file that contains all files part of the feature.
+
++-- feature
+|    +-- feature.json
+|    +-- validate.sh (default)
+|    +-- aquire.sh (default)
```

Comment:
Let's have `aquire.sh` run in a separate build stage to help with caching. (https://github.com/devcontainers/spec/issues/21)

Reply:
nit: "acquire"

Reply:
With other discussions we haven't seen much use with `validate` so we can define them as that features are added in to passes, first all `aquire` scripts are executed and then all `install` ones.

Reply:
Might be worth calling out what is intended to be in these scripts.  If acquire can run as non-root and outputs to a folder, then something needs to be able to install native dependencies intended to do the acquisition... and the acquire script can't do it.  If we're not making that distinction right now, then one entry point may be clearer TBH.

The interesting thing here is the buildpack implementation would have root-requiring steps run *first* and then steps that can land just in a folder second... and nothing after.  I think that's the reverse of what is here. The reason the issue @chrmarti referenced included both a "verify-prereqs.sh" and a "configure.sh" was that you needed to be able to install acquisition dependencies -- e.g. curl or wget -- before you can execute a script.    That has to run before "aquire.sh".  So it was verify-prereqs => acquire => configure. The buildpack model then doesn't have "configure" at all... it more configure => acquire

Reply:
Ok I see I get what you mean. I think this gets into a interesting discussion about exactly how much work we expect/want a feature to do. On one side there is the do everything required to get something to work, on the other is just install what is asked and expect the user to put other features for covering those things. I think the idea is somewhere in between but it gets interesting because I feel like the current features do too many things to not depend in something else. Which is good if they are intended to be completely independent but complicates using multiple features together.

Reply:
The dependencies I'm talking about are those needed by the aquire script to run. e.g. if you're running a build, you need gcc, if you're just downloading something you need curl/wget, if you're cloning something you need git. There's no way someone using a feature can know that up front without reading docs... and we're trying to prevent them from having to do that.

Reply:
One possible way to resolve this if we want to keep two scripts is to reverse the order.

install => acquire

Install focuses on adding anything that requires root access including native OS capabilities, config, updating the home directory, etc.  acquire then focuses on basic acquisition and has a more flexible caching model. 

We should also be sure to call out and set an environment variable for a target path for acquire... for caching to work we'd need to dictate where the install happens.

Reply:
Ok I get what you mean @Chuxel, yeah my idea is that each feature is responsible of validating its os dependencies and installing them. I also like the idea of splitting them for caching purposes.

@chrmarti does that mean that the first script would be expected to run parallel but the second script would run in sequence for when a feature depends on a different one?

Reply:
> To optimize for ease of use for feature authors, we could call the final script run as root in the "run image" install.sh. E.g., a simple feature might just run apt-get install ... in that script.

>
>acquire.sh would run as root in the "build image" and prepare a folder that is later available in the install.sh. E.g., a simple performance-aware feature might just download binaries to that folder and later unpack them to system folders in install.sh. (This allows the download to run in parallel with other acquire.sh scripts.) (https://github.com/devcontainers/spec/issues/21 also has validate-prereqs.sh run as root in the build image and acquire.sh run as regular user in the build image, but I'm not sure we need this distinction.)

>
>To fully optimize performance a feature would do everything in acquire.sh (for parallelization and cachability) and use install.sh only to add symlinks to its files in the final image.

@chrmarti Yeah, though I don't think we should assume `acqure.sh` will have root access in all implementations. It should place content in a specific location that is passed into it that already has privs set. In some ways it would be ideal if it never had root access so it discourages adding content anywhere outside of the folders that will be cached/copied. (There just may not always be a non-root user to do that - so maybe we don't want to force it?)

I think we do need to ensure proper sequencing here however. To describe the intent, those phases were:

1. [Root]: Prepare the (build and run) image with any native OS dependencies - and you could in concept install external .debs here too.
2. [Non-Root]: Acquire, perform local setup/build, and place the result in a prescribed location
3. [Root, Optional for many things] Do any final configuration. This should be very minimal and often optional.  e.g. run ldconfig

Perhaps we should decouple this conversation from the initial spec for features.  Feels like we may want to have a meeting on these nuances.  This can add just `install.sh`, and we decide what the other scripts are for #21 separately.

Reply:
> @chrmarti Yeah, though I don't think we should assume `acqure.sh` will have root access in all implementations. It should place content in a specific location that is passed into it that already has privs set. In some ways it could be ideal if it never had root access so it discourages adding content anywhere outside of the folders that will be cached/copied. (There just may not always be a non-root user to do that - so maybe we don't want to force it?)

Any other implementation would have the ability to run scripts as root when building the image. Having a single script run on the build image will avoid limitations imposed by running one with and one without root access.

> I think we do need to ensure proper sequencing here however. To describe the intent, those phases were:
> 
> 1. [Root]: Prepare the (build and run) image with any native OS dependencies - and you could in concept install external .debs here too.
> 2. [Non-Root]: Acquire, perform local setup/build, and place the result in a prescribed location
> 3. [Root, Optional for many things] Do any final configuration. This should be very minimal and often optional.  e.g. setup and run ldconfig.
> 
> So, from a concurrency perspective, we could do (1) for all features, then concurrently run (2) since it would be writing to separate folders in each case, and finally (3).

Running `acquire.sh` in parallel and then `install.sh` would be a simpler model to get started with.

> Perhaps we should decouple this conversation from the initial spec for features. Feels like we may want to have a meeting on these nuances. This can add just `install.sh`, and we decide what the other scripts are for #21 separately maybe?

Sounds good.

Reply:
> @chrmarti does that mean that the first script would be expected to run parallel but the second script would run in sequence for when a feature depends on a different one?

That would make sense.

Reply:
> Any other implementation would have the ability to run scripts as root when building the image. Having a single script run on the build image will avoid limitations imposed by running one with and one without root access.
>
> ..
>
> Running acquire.sh in parallel and then install.sh would be a simpler model to get started with.

Yeah, unfortunately this would preclude `acquire.sh` from being part of a build step in a buildpack. It runs as UID 1000 (typically a user called `cnb`) in the build container. Anyway, sounds like we're all agreeing that for now just having `install.sh` makes sense while we iron these other aspects out separately.